File: SoftUni - Advanced/Iterators and Comparators/Iterators and Comparators - Exercise/02. Collection/ListyIterator.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;

namespace _2.Collection
{
    public class ListyIterator<T> : IEnumerable<T>
    {
        private List<T> collection;

        private int currentIndex;

        public ListyIterator(params T[] data)
        {
            collection = new List<T>(data);
            currentIndex = 0;
        }
        

        public bool HasNext() => currentIndex < collection.Count - 1;

        public bool Move()
        {
            bool canMove = HasNext();
            if (canMove)
            {
                currentIndex++;
            }

            return canMove;
        }

        public void Print()
        {
            if (collection.Count == 0)
            {
                throw new InvalidOperationException("Invalid Operation!");
            }

            Console.WriteLine($"{collection[currentIndex]}");
        }

        public void PrintAll()
        {
            Console.WriteLine(string.Join(" ", collection));
        }

        public IEnumerator<T> GetEnumerator()
        {
            foreach (T element in collection)
            {
                yield return element;
            }
        }

        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }
}

File: SoftUni - Java Advanced/Iterators and Comparators/Iterators and Comparators - Exercises/_01_ListyIterator/ListyIterator.java
package _01_ListyIterator;

import java.util.ArrayList;
import java.util.List;

public class ListyIterator<T> {
    private List<T> listElements;
    private int currentIndex;

    public ListyIterator(T...elements) {
        this.listElements = List.of(elements);
        this.currentIndex = 0;
    }
    public boolean move() {
        if(hasNext()) {
            this.currentIndex++;
            return true;
        }
        return false;
    }

    public boolean hasNext() {
        return this.currentIndex < this.listElements.size() - 1;
    }
    public void print() {
        if(listElements.isEmpty()) {
            throw new IllegalStateException("Invalid Operation!");
        }
        else {
            System.out.println(this.listElements.get(this.currentIndex));
        }
    }
}
